### PR TITLE
fix NPE bug of URL

### PR DESCRIPTION
**Pull request by wxbty**
What is the purpose of the change

When the org.apache.dubbo.common.URL uses setHost, setPassword, there is an NPE risk of the urlAddress property, because the URL has an empty constructor. For example, when zk pushes the url to the listener, it will call the DefaultServiceInstance.toURL method to construct the URL

    @Override
        public InstanceAddressURL toURL(String protocol) {
            if (instanceAddressURL == null) {
                instanceAddressURL = new InstanceAddressURL(this, serviceMetadata, protocol);
            }
            return instanceAddressURL;
        }

The constructor of InstanceAddressURL here only sets its own properties, which is equivalent to calling the empty constructor of the parent class URL. The constructed URL instance urlAddress property is empty. When someone calls the setHost of this URL, an NPE error will be reported, such as this bug #issue https://github.com/apache/dubbo-spi-extensions/issues/166

The root cause of this error is the dubbo URL design problem, dubbo-spi-extensions did not judge, and then stepped into the trap.

Brief changelog

Verifying this change

Checklist

- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [x] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [x] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Review by AlbumenJ — COMMENTED** [review]
Let `InstanceAddressURL` override these methods would be better

**Comment by chickenlj** [comment]
Yeah, I agree with Albumen

**Comment by wxbty** [comment]
> Let `InstanceAddressURL` override these methods would be better

If urlAddress is empty, does it simply return null? I think we need the necessary NPE check for the URL, which has other subclasses.

**Comment by wxbty** [comment]
hi, @AlbumenJ
Do you think this modification is reasonable? The URL constructor does not use mandatory assignment attributes, so null checking is necessary, the base class should be more robust, and dubbo's extension project dependencies are confusing, and relying on subclasses may still occur frequently NPE.

**Comment by AlbumenJ** [comment]
> > Let `InstanceAddressURL` override these methods would be better
> 
> If urlAddress is empty, does it simply return null? I think we need the necessary NPE check for the URL, which has other subclasses.

In `InstanceAddressURL`, a new URL instance should be created if there are some modifications. The new URL type can be `DubboServiceAddressURL`.

The case of `urlParam` is null is only happends in `InstanceAddressURL`, and override those origin method would fix.

**Comment by wxbty** [comment]
> Do you think this modification is reasonable? The URL constructor does not use mandatory assignment attributes, so null checking is necessary, the base class should be more robust, and dubbo's extension project dependencies are confusing, and relying on subclasses may still occur frequently NPE.

If you think don't need to modify it, I will close it. But still feel that there is a risk of NPE like this, which looks badly designed

**Comment by AlbumenJ** [comment]
> If you think don't need to modify it, I will close it. But still feel that there is a risk of NPE like this, which looks badly designed

Bugfix is needed. 

To make it clarify, I will give some suggest changes for it.

```java
public class InstanceAddressURL {
    @Override
    public URL setHost(String host) {
        // copy current URL into a new DubboServiceAddressURL
        DubboServiceAddressURL copyOf = createDubboServiceAddressURL(this);
      return copyOf.setHost(host);
    }
} 
```

**Comment by AlbumenJ** [comment]
BTW, this issue may be already fixed in #10256.

**Comment by AlbumenJ** [comment]
https://github.com/apache/dubbo/pull/10256/files#diff-8cea6b30b1f89caf0117eb66c7abca7e86e46ecb1f232a8508290ab40ca70ffd

**Comment by AlbumenJ** [comment]
> I noticed this change, it will be fixed in version 3.1, what I want to show is that the parent URL needs to guarantee its NPE check, so that no matter what, Exception won't happen on the path of the URL. If URL defines a new subclass, the author will notice that setHost may return null, and he needs to handle this exception. With the current way of writing, he will think that the URL must be returned. I think, all base class methods should consider null scenarios.

The original solution can indeed solve the NPE problem to a certain extent. (You can reset this PR to the original commit, and I will merge it later.)

In the future, we hope to turn the URL object into an interface at an appropriate time(may be 3.2). The abstract class URL no longer provides attribute storage and implementation, but only provides interface definition.
Based on this, Dubbo's URL system is refactored to avoid the current need for various overriding methods. (Would you please help do this?)

**Comment by wxbty** [comment]
> > I noticed this change, it will be fixed in version 3.1, what I want to show is that the parent URL needs to guarantee its NPE check, so that no matter what, Exception won't happen on the path of the URL. If URL defines a new subclass, the author will notice that setHost may return null, and he needs to handle this exception. With the current way of writing, he will think that the URL must be returned. I think, all base class methods should consider null scenarios.
> 
> The original solution can indeed solve the NPE problem to a certain extent. (You can reset this PR to the original commit, and I will merge it later.)
> 
> In the future, we hope to turn the URL object into an interface at an appropriate time(may be 3.2). The abstract class URL no longer provides attribute storage and implementation, but only provides interface definition. Based on this, Dubbo's URL system is refactored to avoid the current need for various overriding methods. (Would you please help do this?)

Ok, I'll be happy to do it

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 1]
```diff
@@ -499,8 +529,15 @@ public String getPath() {
...
-        return returnURL(newURLAddress);
+        URL url;
+        if (urlAddress == null) {
+            url = new URL();
+            url.setPath(path);
+        } else {
+            URLAddress newURLAddress = urlAddress.setPath(path);
+            url = returnURL(newURLAddress);
+        }
+        return url;
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java:540]:
```suggestion
        if (urlAddress == null) {
            return new ServiceConfigURL(getProtocol(), getHost(), getPort(), path, getParameters());
        } else {
            URLAddress newURLAddress = urlAddress.setPath(path);
            return returnURL(newURLAddress);
        }
```

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 2]
```diff
@@ -458,12 +480,20 @@ public URL setAddress(String address) {
...
+        URL url;
+        if (urlAddress == null) {
+            url = new URL();
+            url.setHost(host);
+            url.setPort(port);
+        } else {
+            URLAddress newURLAddress = urlAddress.setAddress(host, port);
+            url = returnURL(newURLAddress);
+        }
+        return url;
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java:492]:
```suggestion
        if (urlAddress == null) {
            return new ServiceConfigURL(getProtocol(), host, port, getPath(), getParameters());
        } else {
            URLAddress newURLAddress = urlAddress.setAddress(host, port);
            return returnURL(newURLAddress);
        }
```

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 3]
```diff
@@ -435,8 +450,15 @@ public int getPort() {
...
-        return returnURL(newURLAddress);
+        URL url;
+        if (urlAddress == null) {
+            url = new URL();
+            url.setPort(port);
+        } else {
+            URLAddress newURLAddress = urlAddress.setPort(port);
+            url = returnURL(newURLAddress);
+        }
+        return url;
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java:461]:
```suggestion
        if (urlAddress == null) {
            return new ServiceConfigURL(getProtocol(), getHost(), port, getPath(), getParameters());
        } else {
            URLAddress newURLAddress = urlAddress.setPort(port);
            return returnURL(newURLAddress);
        }
```

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 4]
```diff
@@ -425,8 +433,15 @@ public String getHost() {
...
-        return returnURL(newURLAddress);
+        URL url;
+        if (urlAddress == null) {
+            url = new URL();
+            url.setHost(host);
+        } else {
+            URLAddress newURLAddress = urlAddress.setHost(host);
+            url = returnURL(newURLAddress);
+        }
+        return url;
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java:444]:
```suggestion
```suggestion
        if (urlAddress == null) {
            return new ServiceConfigURL(getProtocol(), host, getPort(), getPath(), getParameters());
        } else {
            URLAddress newURLAddress = urlAddress.setHost(host);
            return returnURL(newURLAddress);
        }
```
```

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 5]
```diff
@@ -425,8 +433,13 @@ public String getHost() {
     }
 
     public URL setHost(String host) {
-        URLAddress newURLAddress = urlAddress.setHost(host);
-        return returnURL(newURLAddress);
+```suggestion
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java:436]: remove this breaking line

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 6]
```diff
@@ -367,8 +367,16 @@ public String getPassword() {
...
+
+        URL url;
+        if (urlAddress == null) {
+            url = new URL();
+            url.setPassword(password);
+        } else {
+            URLAddress newURLAddress = urlAddress.setPassword(password);
+            url = returnURL(newURLAddress);
+        }
+        return url;
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java:379]:
```suggestion
        if (urlAddress == null) {
            return new ServiceConfigURL(getProtocol(), getHost(), port, getPath(), getParameters()).setPassword(password);
        } else {
            URLAddress newURLAddress = urlAddress.setPassword(password);
            return returnURL(newURLAddress);
        }
```

Comment (AlbumenJ) [reply dubbo-common/src/main/java/org/apache/dubbo/common/URL.java:379]: BTW, please fix `setUsername`, `setProtocol` and etc

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 7]
```diff
@@ -349,26 +349,38 @@ public String getProtocol() {
     }
 
     public URL setProtocol(String protocol) {
-        URLAddress newURLAddress = urlAddress.setProtocol(protocol);
-        return returnURL(newURLAddress);
+        if (urlAddress == null) {
+            return new ServiceConfigURL(getProtocol(), getHost(), getPort(), getPath(), getParameters());
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java:353]:
```suggestion
            return new ServiceConfigURL(protocol, getHost(), getPort(), getPath(), getParameters());
```

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Review by AlbumenJ — APPROVED** [review]
LGTM

**Pull request merged**